src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,183 @@
+            }
+        }
+
+        /// <summary>
+        /// Add error list entries
+        /// </summary>
+        public void AddEntries(params ErrorListTableEntry[] entries)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            lock (_lockObj)
+            {
+                var sink = _tableSubscription?.TableDataSink;
+
+                if (sink != null)
+                {
+                    sink.AddEntries(entries.ToList(), removeAllEntries: false);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Show error window.
+        /// </summary>
+        public void BringToFront()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            EnsureInitialized();
+
+            // Give the error list focus.
+            var vsErrorList = _errorList as IVsErrorList;
+            vsErrorList?.BringToFront();
+        }
+
+        private void EnsureInitialized()
+        {
+            if (!_initialized)
+            {
+                ThreadHelper.ThrowIfNotOnUIThread();
+
+                _errorList = _serviceProvider.GetService(typeof(SVsErrorList)) as IErrorList;
+                _tableManager = _errorList?.TableControl?.Manager;
+
+                if (_tableManager != null)
+                {
+                    _initialized = true;
+                    _tableManager.AddSource(this);
+                }
+            }
+        }
+
+        private static bool IsNuGetEntry(ITableEntryHandle entry)
+        {
+            object sourceObj;
+            return (entry != null
+                && entry.TryGetValue(StandardTableColumnDefinitions.ErrorSource, out sourceObj)
+                && StringComparer.Ordinal.Equals(ErrorListTableEntry.ErrorSouce, (sourceObj as string)));
+        }
+
+        public void Dispose()
+        {
+            if (_tableManager != null && _initialized)
+            {
+                try
+                {
+                    NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+                    {
+                        await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                        _tableManager.RemoveSource(this);
+                    });
+                }
+                catch (ObjectDisposedException)
+                {
+                    // Ignore disposed exceptions
+                }
+            }
+        }
+
+        /// <summary>
+        /// Holds an ITableDataSink and lock.
+        /// </summary>
+        private class TableSubscription : IDisposable
+        {
+            public ITableDataSink TableDataSink { get; private set; }
+
+            private readonly object _lockObj;
+
+            public TableSubscription(ITableDataSink sink, object lockObj)
+            {
+                TableDataSink = sink;
+                _lockObj = lockObj;
+            }
+
+            public void Dispose()
+            {
+                lock (_lockObj)
```

Comment:
@alpaix this dispose is called by the table manager, it isn't called by NuGet shutting down